test/test_search.rb

```diff
@@ -0,0 +1,96 @@
+Hobbits are small. Hobbits live in Hobbiton.
+One thing Hobbits do not like is Orcs.
+Something a hobbit does like is pastry.
+Some other words.
+EOF
+
+  SEARCH_TEST_LINES = SEARCH_TEST_PAGE.split(/[\n\r]+/)
+
+  setup do
+    @path = cloned_testpath('examples/lotr.git')
+    @wiki = Gollum::Wiki.new(@path, :page_file_dir => 'Rivendell')
+    @wiki.write_page('bar', :markdown, 'bar', commit_details)
+    @wiki.write_page('filename:with:colons', :markdown, '# Filename with colons', commit_details)
+    @wiki.write_page('foo', :markdown, "# File with query in contents and filename\nfoo", commit_details)
+    @wiki.write_page('Hobbit Info', :markdown, SEARCH_TEST_PAGE, commit_details)
+  end
+  
+  test 'search results should be able to return a filename with an embedded colon' do
+    results = @wiki.search('colons')
+    assert_not_nil results
+    assert_equal 'filename:with:colons.md', results.first[:name]
+    assert_equal 2, results.first[:count]
+  end
+
+  test 'search results should make the content/filename search additive' do
+    # There is a file that contains the word 'foo' and is called 'foo', so it should
+    # have a count of 2, not 1...
+    results = @wiki.search('foo')
+    assert_equal 2, results.first[:count]
+  end
+
+  test 'search results should not include files that do not match the query' do
+    results = @wiki.search('foo')
+    assert_equal 1, results.size
+    assert_equal 'foo.md', results.first[:name]
+  end
+
+  test 'search should support multiple search terms ' do
+    results = @wiki.search('foo bar')
+    assert_equal 2, results.size
+    assert_equal 'bar.md', results.first[:name]
+    assert_equal 'foo.md', results.last[:name]
+
+    results = @wiki.search('Hobbits pastry')
+    assert_equal 1, results.size
+    assert_equal 4, results.first[:count]
+    assert_equal SEARCH_TEST_LINES[2], results.first[:context].last
+  end
+
+  test 'search should respect quoted search terms' do
+    results = @wiki.search('like is pastry')
+    hobbit_info_result = results.find {|result| result[:name] == 'Hobbit Info.md'}
+    assert_equal 5, hobbit_info_result[:count]
+    results = @wiki.search('"like is" pastry')
+    assert_equal 3, results.first[:count]
+  end
+
+  test 'search should be case insensitive' do
+    results = @wiki.search('Hobbit')
+    hobbit_info_result = results.find {|result| result[:name] == 'Hobbit Info.md'}
+    assert_equal 6, hobbit_info_result[:count]
+  end
+
+  test 'search should return context for hits' do
+    results = @wiki.search('Hobbit')
+    context = results.first[:context]
+    assert_not_nil context
+    lines = SEARCH_TEST_PAGE.split("\n")
+    assert context.include?(SEARCH_TEST_LINES[0])
+    assert context.include?(SEARCH_TEST_LINES[1])
+    assert context.include?(SEARCH_TEST_LINES[2])
+    assert !context.include?(SEARCH_TEST_LINES[3])
+  end
+
+  test 'search returns escaped search terms on demand' do
+    result, search_terms = @wiki.search('foo|bar', true)
```

Comment:
Are exclusive disjunctions (and conjunctions for that matter) respected?

Reply:
This test should maybe be a bit clearer: what it does is *not* respect the `|` by escaping it. :) That is: we shouldn't let the user attempt to construct a regexp, because that will break our own scanning regex in the adapter.

I think XAND is supported via the `"quote words"` functionality: see the "respect quoted search terms" test. Or were you thinking of something else?

Searching disjunctions in this implementation is limited to this: the search string `foo bar XYZ` will score a hit for each line that contains either `foo`, `bar` or `XYZ`.

If you want to search for files that contain *both* `foo`, and (`bar` XOR `XYZ`), you *could* use the following search string:  `"foo bar" "foo XYZ" "XYZ foo" "bar foo"`

What you can't do is search for files that contain *both* `foo`, and (`bar` XOR `XYZ`). That would be cool, but would make the implementation much more complicated I think (not saying we shouldn't do it -- what do you think?)

Reply:
@bartkamphorst what do you think about this? Is the current implementation intuitive, and should we implement more complex search patterns?

Reply:
First of all, I think the new behavior will be an improvement over the previous behavior. My question pertained mostly to your second example of `foo AND (bar XOR XYZ)`, or even more generally to `(bar XOR foo)` and to `bar AND NOT foo`. The latter case in particular would be a great addition, so that you would be able to search for "Hobbit AND NOT Bilbo". 😉 I don't know how feasible any of those options are though.

Reply:
I agree they would be cool, but constructing regexes based on user input like that feels like re-inventing the wheel a bit. It would be great if there were a generic solution for generating regexes out of user search strings like that -- or alternatively, if we could just let the user type regexes.

Reply:
I've been looking but I haven't found anything out of the box that does what we want without too much overhead. Here's [one approach](https://www.sitepoint.com/community/t/need-help-converting-user-friendly-search-string-to-regex/85747) though that might serve as inspiration.

Reply:
> or alternatively, if we could just let the user type regexes.

There's some security issues with this approach, for example DOS attacks if the expression takes long to resolve.

I'm surprised there's no standard way of dealing with this other than using full-blown search libraries.

Reply:
See https://gist.github.com/dometto/c9ffadfaacd8f50c31d69412a65ba64f for most advanced implementation (supporting NOT and AND) I can get to work without implementing a full-fledged parser. Note that I couldn't get disjunctions within search terms to work :(

Reply:
I have changed my mind about this. I think we should just start with the scope of the current PR, with context, multiple keyword search and exact matches. 👍